File/Directory Entries API: Make tests Edge friendly

inexorabletash opened the pull request:
Edge doesn't yet implement getFile()/getDirectory()/getParent() so don't make unrelated tests assume the existence of those methods. Add a helper to get child entries via enumeration (supported everywhere) and also split up the test files by method.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

inexorabletash committed:
Reduce dependencies on APIs by tests

inexorabletash committed:
Tweak title

inexorabletash commented:
@aliams @cdumez @mkruisselbrink @pwnall - could one of you take a look at these changes?

w3c-bots commented:
# Build PASSED

Started: 2017-10-10 21:42:01
Finished: 2017-10-10 21:49:07


<h2>Failing Jobs</h2>
<ul>

<li>chrome:unstable</li>

<li>safari:10.0</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/19300)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/286245265)

mkruisselbrink commented on entries-api/support.js:
```diff
@@ -10,15 +10,16 @@ window.addEventListener('DOMContentLoaded', e => {
     e.preventDefault();
-    for (const item of e.dataTransfer.items) {
+    for (var i = 0; i < e.dataTransfer.items.length; ++i) {
+      var item = e.dataTransfer.items[i];
```
nit: const?

inexorabletash replied:
done

mkruisselbrink commented on entries-api/support.js:
```diff
@@ -10,15 +10,16 @@ window.addEventListener('DOMContentLoaded', e => {
   elem.addEventListener('drop', e => {
     e.preventDefault();
-    for (const item of e.dataTransfer.items) {
+    for (var i = 0; i < e.dataTransfer.items.length; ++i) {
```
nit: this also doesn't have to be var, right?

inexorabletash replied:
Made `const`.

mkruisselbrink commented on entries-api/support.js:
```diff
@@ -89,3 +87,43 @@ const FILE_PATHS = [
+    let doBatch = () => {
+      reader.readEntries(entries => {
+        if (entries.length > 0) {
+          entries.forEach(e => result.push(e));
```
nit: isn't this just Array.prototype.push.apply(result, entries)? Or result = result.concat(entries); Although I guess neither of those is necessarily cleaner/simpler than what you have.

inexorabletash replied:
All correct. Leaving as-is, though. I could make all those `let`s into `const`s though.

mkruisselbrink commented on entries-api/support.js:
```diff
@@ -89,3 +87,43 @@ const FILE_PATHS = [
+
+function getChildEntry(dirEntry, name, callback) {
+  getEntriesAsPromise(dirEntry)
+    .then(function(entries) {
```
nit: entries => {} might be more consistent with the rest of these tests?

inexorabletash replied:
Thanks, meant to fix this and forgot. Done!

mkruisselbrink commented on entries-api/support.js:
```diff
@@ -89,3 +87,43 @@ const FILE_PATHS = [
+function getChildEntry(dirEntry, name, callback) {
+  getEntriesAsPromise(dirEntry)
+    .then(function(entries) {
+      return entries.filter(function(entry) {
```
nit: here too, => rather than function?

inexorabletash replied:
done

mkruisselbrink commented on entries-api/filesystementry-getParent-manual.html:
```diff
@@ -0,0 +1,27 @@
+
+<script>
+entry_test((t, entry) => {
+  assert_idl_attribute(entry, 'getParent', 'FileSystemEntry has filesystem attribute');
```
(I know this is just copied from before) getParent() rather than filesystem, right?

inexorabletash replied:
D'oh, nice catch. Fixed.

mkruisselbrink commented on entries-api/support.js:
```diff
@@ -89,3 +87,43 @@ const FILE_PATHS = [
+        return entry.name === name;
+      })[0];
+    })
+    .then(callback);
```
Shouldn't you handle the promise rejecting to properly make whatever test called getChildEntry fail? Of course you might need to pass in the Test to getChildEntry for that to be possible.

inexorabletash replied:
... or update all the tests to be `promise_test` based. I'll add an `errback`

aliams commented:
Does the name consistency and object identity test need to use getParent in filesystem-manual.html?

aliams commented:
Also, we apparently don't support getFile as well.

inexorabletash committed:
review feedback

inexorabletash committed:
Avoid getParent in filesystem test

inexorabletash commented:
> Does the name consistency and object identity test need to use getParent in filesystem-manual.html?

Good point - I was able to switch that to the `getChildEntry()` helper as well (which uses DirectoryReader)

> Also, we apparently don't support getFile as well.

Yeah, I've sequestered getParent()/getFile()/getDirectory() into specific tests (and the IDL ones). The only exception is errors-manual.html which is verifying all the DOMException details and I think I need getFile for that one.

aliams commented:
Thanks @inexorabletash! Looking over the tests now, the pass rate for what's actually needed to accomplish the directory upload scenario is much better now for Edge.

inexorabletash commented:
Thanks for confirming, @aliams !

And since github's workflow is a bit sketchy: @mkruisselbrink - I've addressed your feedback and pushed an update to the branch if you want to take another look
